File: app/src/main/java/ch/bailu/aat/gpx/attributes/GpxAttributesStatic.java
package ch.bailu.aat.gpx.attributes;


import android.support.annotation.NonNull;

public class GpxAttributesStatic extends GpxAttributes {


    public static class Tag implements Comparable<Tag> {
        public Tag(Tag keyValue) {
            this(keyValue.key, keyValue.value);
        }

        public Tag(int k, String v) {
            key=k;
            value=v;
        }

        public final int key;
        public final String value;


        @Override
        public int compareTo(@NonNull Tag another) {
            return Integer.compare(key, another.key);
        }
    }

    private Tag[] tagList;


    public GpxAttributesStatic() {
        this(new Tag[]{});
    }

    public GpxAttributesStatic(Tag[] attr) {
        tagList=attr;
    }


    @Override
    public String get(int key) {
        int index = getIndex(key);

        if (index==size()) return NULL_VALUE;

        return tagList[index].value;
    }


    @Override
    public boolean hasKey(int key) {
        return getIndex(key) < size();
    }

    @Override
    public void put(int key, String value) {
        int index = getIndex(key);

        if (index == size()) {
            Tag[] newTagList = new Tag[size()+1];

            System.arraycopy(tagList, 0, newTagList, 0, tagList.length);
            tagList=newTagList;
        }

        tagList[index] = new Tag(key, value);
    }


    private int getIndex(int key) {
        for (int i=0; i<size(); i++) {
            if (tagList[i].key == key) return i;
        }
        return size();
    }


    @Override
    public int size() {
        return tagList.length;
    }

    @Override
    public String getAt(int i) {
        return tagList[i].value;
    }

    @Override
    public int getKeyAt(int i) {
        return tagList[i].key;
    }

/*
    @Override
    public void remove(String key) {
        int index = getIndex(key);

        if (index < size()) {
            Tag[] newTagList = new Tag[size()-1];

            int newIndex=0;
            for (int oldIndex=0; oldIndex<size(); oldIndex++) {
                if (oldIndex != index) {
                    newTagList[newIndex] = new Tag(tagList[oldIndex]);
                    newIndex++;
                }
            }
            tagList=newTagList;
        }
    }
*/
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/attributes/GpxAttributesStatic.kt
package ch.bailu.aat_lib.gpx.attributes

class GpxAttributesStatic(private var tagList: Array<Tag> = arrayOf()) :
    GpxAttributes() {
    class Tag(val key: Int, val value: String) : Comparable<Tag> {
        override fun compareTo( other: Tag): Int {
            return key.compareTo(other.key)
        }
    }

    override fun get(keyIndex: Int): String {
        val index = getIndex(keyIndex)
        return if (index == size()) NULL_VALUE else tagList[index].value
    }

    override fun hasKey(keyIndex: Int): Boolean {
        return getIndex(keyIndex) < size()
    }

    override fun put(keyIndex: Int, value: String) {
        val index = getIndex(keyIndex)
        val tag = Tag(keyIndex, value)

        if (index == size()) {
            tagList = arrayOf(*tagList, tag)
        } else {
            tagList[index] = tag
        }
    }

    private fun getIndex(key: Int): Int {
        for (i in 0 until size()) {
            if (tagList[i].key == key) return i
        }
        return size()
    }

    override fun size(): Int {
        return tagList.size
    }

    override fun getAt(index: Int): String {
        return tagList[index].value
    }

    override fun getKeyAt(index: Int): Int {
        return tagList[index].key
    }
}
